starter-code/src/functions-and-arrays.js

```diff
@@ -72,30 +150,406 @@ const wordsCount = [
+    17,
+    77,
+    4,
+    89,
+    55,
+    40,
+  ],
   [4, 52, 8, 83, 97, 35, 99, 16, 7, 97, 57, 32, 16, 26, 26, 79, 33, 27, 98, 66],
-  [88, 36, 68, 87, 57, 62, 20, 72, 3, 46, 33, 67, 46, 55, 12, 32, 63, 93, 53, 69],
-  [4, 42, 16, 73, 38, 25, 39, 11, 24, 94, 72, 18, 8, 46, 29, 32, 40, 62, 76, 36],
-  [20, 69, 36, 41, 72, 30, 23, 88, 34, 62, 99, 69, 82, 67, 59, 85, 74, 4, 36, 16],
-  [20, 73, 35, 29, 78, 31, 90, 1, 74, 31, 49, 71, 48, 86, 81, 16, 23, 57, 5, 54],
-  [1, 70, 54, 71, 83, 51, 54, 69, 16, 92, 33, 48, 61, 43, 52, 1, 89, 19, 67, 48]
+  [
+    88,
+    36,
+    68,
+    87,
+    57,
+    62,
+    20,
+    72,
+    3,
+    46,
+    33,
+    67,
+    46,
+    55,
+    12,
+    32,
+    63,
+    93,
+    53,
+    69,
+  ],
+  [
+    4,
+    42,
+    16,
+    73,
+    38,
+    25,
+    39,
+    11,
+    24,
+    94,
+    72,
+    18,
+    8,
+    46,
+    29,
+    32,
+    40,
+    62,
+    76,
+    36,
+  ],
+  [
+    20,
+    69,
+    36,
+    41,
+    72,
+    30,
+    23,
+    88,
+    34,
+    62,
+    99,
+    69,
+    82,
+    67,
+    59,
+    85,
+    74,
+    4,
+    36,
+    16,
+  ],
+  [
+    20,
+    73,
+    35,
+    29,
+    78,
+    31,
+    90,
+    1,
+    74,
+    31,
+    49,
+    71,
+    48,
+    86,
+    81,
+    16,
+    23,
+    57,
+    5,
+    54,
+  ],
+  [
+    1,
+    70,
+    54,
+    71,
+    83,
+    51,
+    54,
+    69,
+    16,
+    92,
+    33,
+    48,
+    61,
+    43,
+    52,
+    1,
+    89,
+    19,
+    67,
+    48,
+  ],
 ];
+function greatestProduct(matrix) {
+  let value = 0,
+    greatest = 0;
+  if (matrix === undefined) {
+    return null;
+  }
+  // Find the point that we want to check
+  // What are the values up, down, left, and right from that point.
+  // For each value found, multiply
+  // Check the next point
+  for (let i = 0; i < matrix.length; i++) {
+    let row = matrix[i].length;
+    let result = 0;
+    // console.log(row);
+    for (let j = 0; j < row; j++) {
+      let values = { up: [], down: [], left: [], right: [] };
+      // Check up
+      values.up.push(matrix[i][j]);
+      for (adj = 1; adj < 4; adj++) {
+        if (i - adj >= 0) {
+          values.up.push(matrix[i - adj][j]);
+        }
+      }
+      // Check down
+      values.down.push(matrix[i][j]);
+      for (adj = 1; adj < 4; adj++) {
+        if (i + adj < matrix.length) {
+          values.down.push(matrix[i + adj][j]);
+        }
+      }
+      // Check left
+      values.left.push(matrix[i][j]);
+      for (adj = 1; adj < 4; adj++) {
+        if (j - adj >= 0) {
+          values.left.push(matrix[i][j - adj]);
+        }
+      }
+      // Check right
+      values.right.push(matrix[i][j]);
+      for (adj = 1; adj < 4; adj++) {
+        if (j + adj < row) {
+          values.right.push(matrix[i][j + adj]);
+        }
+      }
+
+      // Check what values can be multiplied
+
+      value = 0;
+      for (let key in values) {
+        for (let num = 0; num < values[key].length; num++) {
```

Comment:
There's no issue using another nested loop for a problem like this, but nested for loops often perform very slow. Be mindful of performance in the future.